apps/perl/5.18.2/lib/5.18.2/man/man3/JSON.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "JSON 3"
.TH JSON 3 "2013-10-31" "perl v5.18.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
JSON \- JSON (JavaScript Object Notation) encoder/decoder
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& use JSON; # imports encode_json, decode_json, to_json and from_json.
\& 
\& # simple and fast interfaces (expect/generate UTF\-8)
\& 
\& $utf8_encoded_json_text = encode_json $perl_hash_or_arrayref;
\& $perl_hash_or_arrayref  = decode_json $utf8_encoded_json_text;
\& 
\& # OO\-interface
\& 
\& $json = JSON\->new\->allow_nonref;
\& 
\& $json_text   = $json\->encode( $perl_scalar );
\& $perl_scalar = $json\->decode( $json_text );
\& 
\& $pretty_printed = $json\->pretty\->encode( $perl_scalar ); # pretty\-printing
\& 
\& # If you want to use PP only support features, call with \*(Aq\-support_by_pp\*(Aq
\& # When XS unsupported feature is enable, using PP (de|en)code instead of XS ones.
\& 
\& use JSON \-support_by_pp;
\& 
\& # option\-acceptable interfaces (expect/generate UNICODE by default)
\& 
\& $json_text   = to_json( $perl_scalar, { ascii => 1, pretty => 1 } );
\& $perl_scalar = from_json( $json_text, { utf8  => 1 } );
\& 
\& # Between (en|de)code_json and (to|from)_json, if you want to write
\& # a code which communicates to an outer world (encoded in UTF\-8),
\& # recommend to use (en|de)code_json.
.Ve
.SH "VERSION"
.IX Header "VERSION"
.Vb 1
\&    2.90
.Ve
.PP
This version is compatible with \s-1JSON::XS \s0\fB2.34\fR and later.
(Not yet compatble to \s-1JSON::XS \s0\fB3.0x\fR.)
.SH "NOTE"
.IX Header "NOTE"
\&\s-1JSON::PP\s0 was earlier included in the \f(CW\*(C`JSON\*(C'\fR distribution, but
has since Perl 5.14 been a core module. For this reason,
\&\s-1JSON::PP\s0 was removed from the \s-1JSON\s0 distribution and can now
be found also in the Perl5 repository at
.IP "\(bu" 4
<http://perl5.git.perl.org/perl.git>
.PP
(The newest \s-1JSON::PP\s0 version still exists in \s-1CPAN.\s0)
.PP
Instead, the \f(CW\*(C`JSON\*(C'\fR distribution will include JSON::backportPP
for backwards computability. \s-1JSON\s0.pm should thus work as it did
before.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.Vb 10
\& *************************** CAUTION **************************************
\& *                                                                        *
\& * INCOMPATIBLE CHANGE (JSON::XS version 2.90)                            *
\& *                                                                        *
\& * JSON.pm had patched JSON::XS::Boolean and JSON::PP::Boolean internally *
\& * on loading time for making these modules inherit JSON::Boolean.        *
\& * But since JSON::XS v3.0 it use Types::Serialiser as boolean class.     *
\& * Then now JSON.pm breaks boolean classe overload features and           *
\& * \-support_by_pp if JSON::XS v3.0 or later is installed.                 *
\& *                                                                        *
\& * JSON::true and JSON::false returned JSON::Boolean objects.             *
\& * For workaround, they return JSON::PP::Boolean objects in this version. *
\& *                                                                        *
\& *     isa_ok(JSON::true, \*(AqJSON::PP::Boolean\*(Aq);                           *
\& *                                                                        *
\& * And it discards a feature:                                             *
\& *                                                                        *
\& *     ok(JSON::true eq \*(Aqtrue\*(Aq);                                          *
\& *                                                                        *
\& * In other word, JSON::PP::Boolean overload numeric only.                *
\& *                                                                        *
\& *     ok( JSON::true == 1 );                                             *
\& *                                                                        *
\& **************************************************************************
\&
\& ************************** CAUTION ********************************
\& * This is \*(AqJSON module version 2\*(Aq and there are many differences  *
\& * to version 1.xx                                                 *
\& * Please check your applications using old version.              *
\& *   See to \*(AqINCOMPATIBLE CHANGES TO OLD VERSION\*(Aq                  *
\& *******************************************************************
.Ve
.PP
\&\s-1JSON \s0(JavaScript Object Notation) is a simple data format.
See to <http://www.json.org/> and \f(CW\*(C`RFC4627\*(C'\fR(<http://www.ietf.org/rfc/rfc4627.txt>).
.PP
This module converts Perl data structures to \s-1JSON\s0 and vice versa using either
\&\s-1JSON::XS\s0 or \s-1JSON::PP\s0.
.PP
\&\s-1JSON::XS\s0 is the fastest and most proper \s-1JSON\s0 module on \s-1CPAN\s0 which must be
compiled and installed in your environment.
\&\s-1JSON::PP\s0 is a pure-Perl module which is bundled in this distribution and
has a strong compatibility to \s-1JSON::XS.\s0
.PP
This module try to use \s-1JSON::XS\s0 by default and fail to it, use \s-1JSON::PP\s0 instead.
So its features completely depend on \s-1JSON::XS\s0 or \s-1JSON::PP.\s0
.PP
See to \*(L"\s-1BACKEND MODULE DECISION\*(R"\s0.
.PP
To distinguish the module name '\s-1JSON\s0' and the format type \s-1JSON,\s0
the former is quoted by C<> (its results vary with your using media),
and the latter is left just as it is.
.PP
Module name : \f(CW\*(C`JSON\*(C'\fR
.PP
Format type : \s-1JSON\s0
.SS "\s-1FEATURES\s0"
.IX Subsection "FEATURES"
.IP "\(bu" 4
correct unicode handling
.Sp
This module (i.e. backend modules) knows how to handle Unicode, documents
how and when it does so, and even documents what \*(L"correct\*(R" means.
.Sp
Even though there are limitations, this feature is available since Perl version 5.6.
.Sp
\&\s-1JSON::XS\s0 requires Perl 5.8.2 (but works correctly in 5.8.8 or later), so in older versions
\&\f(CW\*(C`JSON\*(C'\fR should call \s-1JSON::PP\s0 as the backend which can be used since Perl 5.005.
.Sp
With Perl 5.8.x \s-1JSON::PP\s0 works, but from 5.8.0 to 5.8.2, because of a Perl side problem,
\&\s-1JSON::PP\s0 works slower in the versions. And in 5.005, the Unicode handling is not available.
See to \*(L"\s-1UNICODE HANDLING ON PERLS\*(R"\s0 in \s-1JSON::PP\s0 for more information.
.Sp
See also to \*(L"A \s-1FEW NOTES ON UNICODE AND PERL\*(R"\s0 in \s-1JSON::XS\s0
and \*(L"\s-1ENCODING/CODESET_FLAG_NOTES\*(R"\s0 in \s-1JSON::XS\s0.
.IP "\(bu" 4
round-trip integrity
.Sp
When you serialise a perl data structure using only data types supported
by \s-1JSON\s0 and Perl, the deserialised data structure is identical on the Perl
level. (e.g. the string \*(L"2.0\*(R" doesn't suddenly become \*(L"2\*(R" just because
it looks like a number). There \fIare\fR minor exceptions to this, read the
\&\*(L"\s-1MAPPING\*(R"\s0 section below to learn about those.
.IP "\(bu" 4
strict checking of \s-1JSON\s0 correctness
.Sp
There is no guessing, no generating of illegal \s-1JSON\s0 texts by default,
and only \s-1JSON\s0 is accepted as input by default (the latter is a security
feature).
.Sp
See to \*(L"\s-1FEATURES\*(R"\s0 in \s-1JSON::XS\s0 and \*(L"\s-1FEATURES\*(R"\s0 in \s-1JSON::PP\s0.
.IP "\(bu" 4
fast
.Sp
This module returns a \s-1JSON::XS\s0 object itself if available.
Compared to other \s-1JSON\s0 modules and other serialisers such as Storable,
\&\s-1JSON::XS\s0 usually compares favorably in terms of speed, too.
.Sp
If not available, \f(CW\*(C`JSON\*(C'\fR returns a \s-1JSON::PP\s0 object instead of \s-1JSON::XS\s0 and
it is very slow as pure-Perl.
.IP "\(bu" 4
simple to use
.Sp
This module has both a simple functional interface as well as an
object oriented interface interface.
.IP "\(bu" 4
reasonably versatile output formats
.Sp
You can choose between the most compact guaranteed-single-line format possible
(nice for simple line-based protocols), a pure-ASCII format (for when your transport
is not 8\-bit clean, still supports the whole Unicode range), or a pretty-printed
format (for when you want to read that stuff). Or you can combine those features
in whatever way you like.
.SH "FUNCTIONAL INTERFACE"
.IX Header "FUNCTIONAL INTERFACE"
Some documents are copied and modified from \*(L"\s-1FUNCTIONAL INTERFACE\*(R"\s0 in \s-1JSON::XS\s0.
\&\f(CW\*(C`to_json\*(C'\fR and \f(CW\*(C`from_json\*(C'\fR are additional functions.
.SS "encode_json"
.IX Subsection "encode_json"
.Vb 1
\&    $json_text = encode_json $perl_scalar
.Ve
.PP
Converts the given Perl data structure to a \s-1UTF\-8\s0 encoded, binary string.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&    $json_text = JSON\->new\->utf8\->encode($perl_scalar)
.Ve
.SS "decode_json"
.IX Subsection "decode_json"
.Vb 1
\&    $perl_scalar = decode_json $json_text
.Ve
.PP
The opposite of \f(CW\*(C`encode_json\*(C'\fR: expects an \s-1UTF\-8 \s0(binary) string and tries
to parse that as an \s-1UTF\-8\s0 encoded \s-1JSON\s0 text, returning the resulting
reference.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&    $perl_scalar = JSON\->new\->utf8\->decode($json_text)
.Ve
.SS "to_json"
.IX Subsection "to_json"
.Vb 1
\&   $json_text = to_json($perl_scalar)
.Ve
.PP
Converts the given Perl data structure to a json string.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&   $json_text = JSON\->new\->encode($perl_scalar)
.Ve
.PP
Takes a hash reference as the second.
.PP
.Vb 1
\&   $json_text = to_json($perl_scalar, $flag_hashref)
.Ve
.PP
So,
.PP
.Vb 1
\&   $json_text = to_json($perl_scalar, {utf8 => 1, pretty => 1})
.Ve
.PP
equivalent to:
.PP
.Vb 1
\&   $json_text = JSON\->new\->utf8(1)\->pretty(1)\->encode($perl_scalar)
.Ve
.PP
If you want to write a modern perl code which communicates to outer world,
you should use \f(CW\*(C`encode_json\*(C'\fR (supposed that \s-1JSON\s0 data are encoded in \s-1UTF\-8\s0).
.SS "from_json"
.IX Subsection "from_json"
.Vb 1
\&   $perl_scalar = from_json($json_text)
.Ve
.PP
The opposite of \f(CW\*(C`to_json\*(C'\fR: expects a json string and tries
to parse it, returning the resulting reference.
.PP
This function call is functionally identical to:
.PP
.Vb 1
\&    $perl_scalar = JSON\->decode($json_text)
.Ve
.PP
Takes a hash reference as the second.
.PP
.Vb 1
\&    $perl_scalar = from_json($json_text, $flag_hashref)
.Ve
.PP
So,
.PP
.Vb 1
\&    $perl_scalar = from_json($json_text, {utf8 => 1})
.Ve
.PP
equivalent to:
.PP
.Vb 1
\&    $perl_scalar = JSON\->new\->utf8(1)\->decode($json_text)
.Ve
.PP
If you want to write a modern perl code which communicates to outer world,
you should use \f(CW\*(C`decode_json\*(C'\fR (supposed that \s-1JSON\s0 data are encoded in \s-1UTF\-8\s0).
.SS "JSON::is_bool"
.IX Subsection "JSON::is_bool"
.Vb 1
\&    $is_boolean = JSON::is_bool($scalar)
.Ve
.PP
Returns true if the passed scalar represents either JSON::true or
JSON::false, two constants that act like \f(CW1\fR and \f(CW0\fR respectively
and are also used to represent \s-1JSON \s0\f(CW\*(C`true\*(C'\fR and \f(CW\*(C`false\*(C'\fR in Perl strings.
.SS "JSON::true"
.IX Subsection "JSON::true"
Returns \s-1JSON\s0 true value which is blessed object.
It \f(CW\*(C`isa\*(C'\fR JSON::Boolean object.
.SS "JSON::false"
.IX Subsection "JSON::false"
Returns \s-1JSON\s0 false value which is blessed object.
It \f(CW\*(C`isa\*(C'\fR JSON::Boolean object.
.SS "JSON::null"
.IX Subsection "JSON::null"
Returns \f(CW\*(C`undef\*(C'\fR.
.PP
See \s-1MAPPING\s0, below, for more information on how \s-1JSON\s0 values are mapped to
Perl.
.SH "HOW DO I DECODE A DATA FROM OUTER AND ENCODE TO OUTER"
.IX Header "HOW DO I DECODE A DATA FROM OUTER AND ENCODE TO OUTER"
This section supposes that your perl version is 5.8 or later.
.PP
If you know a \s-1JSON\s0 text from an outer world \- a network, a file content, and so on,
is encoded in \s-1UTF\-8,\s0 you should use \f(CW\*(C`decode_json\*(C'\fR or \f(CW\*(C`JSON\*(C'\fR module object
with \f(CW\*(C`utf8\*(C'\fR enable. And the decoded result will contain \s-1UNICODE\s0 characters.
.PP
.Vb 4
\&  # from network
\&  my $json        = JSON\->new\->utf8;
\&  my $json_text   = CGI\->new\->param( \*(Aqjson_data\*(Aq );
\&  my $perl_scalar = $json\->decode( $json_text );
\&  
\&  # from file content
\&  local $/;
\&  open( my $fh, \*(Aq<\*(Aq, \*(Aqjson.data\*(Aq );
\&  $json_text   = <$fh>;
\&  $perl_scalar = decode_json( $json_text );
.Ve
.PP
If an outer data is not encoded in \s-1UTF\-8,\s0 firstly you should \f(CW\*(C`decode\*(C'\fR it.
.PP
.Vb 5
\&  use Encode;
\&  local $/;
\&  open( my $fh, \*(Aq<\*(Aq, \*(Aqjson.data\*(Aq );
\&  my $encoding = \*(Aqcp932\*(Aq;
\&  my $unicode_json_text = decode( $encoding, <$fh> ); # UNICODE
\&  
\&  # or you can write the below code.
\&  #
\&  # open( my $fh, "<:encoding($encoding)", \*(Aqjson.data\*(Aq );
\&  # $unicode_json_text = <$fh>;
.Ve
.PP
In this case, \f(CW$unicode_json_text\fR is of course \s-1UNICODE\s0 string.
So you \fBcannot\fR use \f(CW\*(C`decode_json\*(C'\fR nor \f(CW\*(C`JSON\*(C'\fR module object with \f(CW\*(C`utf8\*(C'\fR enable.

lib/Conf/Json.pm
use MooseX::Declare;
use Method::Signatures::Simple;

=head2

	PACKAGE		Conf::Json

    PURPOSE
    
        1. READ JSON ARRAY-FORMAT CONFIGURATION FILES (PRETTY FORMAT)

		2. ADD/EDIT/REMOVE ENTRIES
		
		3. WRITE TO OUTFILE PRESERVING ORDER OF KEYS
		
=cut


class Conf::Json with (Conf, Agua::Common::Logger, Agua::Common::Util) {

use Data::Dumper;
use JSON;

has 'backup' 	=>	(	is	=>	'rw',	isa	=>	'Bool', default	=>	0	);
has 'inputfile' =>	(	is	=>	'rw',	isa	=>	'Str'	);
has 'outputfile' => (	is	=>	'rw',	isa	=>	'Str|Undef'	);
has 'ignore' 	=>	(	is	=>	'rw',	isa	=>	'Str',	default	=>	"#;\\[?"	);
has 'spacer' 	=>	(	is	=>	'rw',	isa	=>	'Str',	default	=>	"\\t=\\s"	);
has 'separator'	=>	(	is	=>	'rw',	isa	=> 	'Str',	default	=>	"="		);
has 'comment'	=>	(	is	=>	'rw',	isa	=> 	'Str',	default	=>	"#"		);
has 'sections'	=>	(	is 	=>	'rw', 	isa =>	'ArrayRef|Undef' );
has 'order'		=>	(	is 	=>	'rw', 	isa =>	'ArrayRef' );
has 'jsonparser'	=> ( isa => 'JSON', is => 'rw', lazy => 1, builder => "setJsonParser" );

####///}}} 
=head2

	SUBROUTINE 		read
	
	PURPOSE
	
		1. IF NO section IS PROVIDED, RETURN A HASH OF {section => {KEY-VALUE PAIRS}} PAIRS
		
		2. IF NO key IS PROVIDED, RETURN A HASH OF KEY-VALUE PAIRS FOR THE section
		
		3. RETURN THE KEY-VALUE PAIR FOR THE section AND key
		
		4. RETURN IF THE GIVEN section IS NOT IN THE ini FILE
		
		5. RETURN IF THE GIVEN key IS NOT IN THE ini FILE

=cut

method read ($inputfile) {
	$inputfile = $self->inputfile() if not defined $inputfile;
	$self->inputfile($inputfile) if defined $inputfile;
	$inputfile = $self->outputfile() if not defined $inputfile;
	$self->logNote("inputfile", $inputfile);
	
	return if not defined $inputfile;
	#$self->logDebug("inputfile ", $inputfile);
	
	#### SKIP READ IF DIRECTORY ABSENT
	my $sections = [];

	#### SKIP READ IF FILE ABSENT
	$self->logNote("not -f $inputfile. Returning sections: ") if not -f $inputfile;
	return $sections if not -f $inputfile;

	#### GET ORDER OF SECTIONS
	my $order = $self->getSectionOrder($inputfile);
	
	#### PARSE JSON DATA
	my $data = $self->parseJson($inputfile);
	
	#### SET SECTIONS
	foreach my $key ( @$order ) {
		push @$sections, {	$key	=>	$data->{$key} };
	}
	
	#print Dumper $sections;
	$self->sections($sections);
	
	return $sections;
}

method getSectionOrder ($inputfile) {
	#$self->logDebug("inputfile", $inputfile);
	my $contents = $self->fileContents($inputfile);
	#$self->logDebug("");
	my $lines;
	@$lines = split "\n", $contents;
	my $order = [];
	my $counter = 0;

	my $line;
	while ( $counter < scalar(@$lines) ) {
		$line = $$lines[$counter];
		$counter++;

		#$self->logDebug("IN WHILE LOOP line $counter", $line);

		#### IF LINE HAS A KEY, STORE IT AND PUSH A CLOSER
		if ( $line =~ /^\s*"([^".]+)"\s*:\s*(\{|\[)\s*$/ ) {
			#$self->logDebug("NESTED pushing to order: $1 IN line", $line);
			push @$order, $1;
			my $opener = $2;
			my $closer = "]";
			$closer = "}" if $opener eq "{";

			#### SKIP ANY SUBLEVELS
			($counter, $lines) = $self->skipSubLevels($counter, $lines, $closer);
		}
		elsif ( $line =~ /^\s*"([^".]+)"\s*:\s*/ ) {
			#$self->logDebug("SIMPLE pushing to order: $1 IN line", $line);
			push @$order, $1;
		}
	}
	$self->order($order);

	return $order;
}

method skipSubLevels ($counter, $lines, $closer ) {	
	
	#### KEEP STACK OF NESTED CLOSERS
	my $stack = [];
	push @$stack, $closer;

	my $line = '';
	$counter--;
	while ( defined $line and scalar(@$stack) ) {
		$counter++;
		$line = $$lines[$counter];
		#$self->logDebug("LOOKING FOR CLOSER '$closer' INSIDE line $counter", $line);

		if ( $line =~ /^\s*"[^".]+"\s*:\s*(\{|\[)\s*$/
			or $line =~ /^\s*(\{|\[)\s*$/ ) {
			my $newopener = $1;
			my $newcloser = "]";
			$newcloser = "}" if $newopener eq "{";
			push @$stack, $newcloser;
			$closer = $newcloser;
			#$self->logDebug("NEW CLOSER '$closer' IN line", $line);
			#$self->logDebug("NEW STACK: @$stack");
		}
		elsif ( $line =~ /$closer,?\s*$/ ) {
			my $popped = pop @$stack;
			$closer = $$stack[scalar(@$stack) - 1] if @$stack;
			$closer = "UNDEFINED" if not @$stack;
			#$self->logDebug("popped closer: '$popped' in line $counter", $line);
			#$self->logDebug("NEW STACK: @$stack");
		}
		elsif ( $line =~ /^\s*(\{|\[)\s*$/ ) {
			my $newopener = $1;
			my $newcloser = "]";
			$newcloser = "}" if $newopener eq "{";
			push @$stack, $newcloser;
			$closer = $newcloser;
			#$self->logDebug("NEW CLOSER '$closer' IN line", $line);
			#$self->logDebug("NEW STACK: @$stack");
		}
	}

	return ($counter, $lines)
}

method copy ($inputfile, $outputfile) {
	$self->logDebug("Conf::copy(file)");
	$self->logDebug("inputfile", $inputfile);
	$self->logDebug("outputfile", $outputfile);
	$self->read($inputfile);
	$self->write($outputfile);	
}

method write ($file) {
	$self->logNote("file", $file) if defined $file;
	$file = $self->outputfile() if not defined $file;	
	$file = $self->inputfile() if not defined $file;	
	#$self->logNote("file", $file);
	my $sections = $self->_getSections();

	#### SANITY CHECK
	$self->logError("Conf::Json::write    file not defined") if not defined $file;
	$self->logError("Conf::Json::write    sections not defined") if not defined $sections;
	
	#### KEEP COPIES OF FILES IF backup DEFINED
	$self->makeBackup($file) if $self->backup();

	#### CREATE FILE DIR IF NOT EXISTS	
	my ($dir) = $file =~ /^(.+)\/([^\/]+)$/;
	File::Path::mkpath($dir) if not -d $dir;
	$self->logError("Conf::Json::write    Can't create dir: $dir") if not -d $dir;

	my $jsonparser = $self->getJsonParser();

	my $output = "{\n";
	foreach my $section ( @$sections )
	{
		my $json = $jsonparser->pretty->indent->encode($section);
		$json =~ s/^\{\n//;
		$json =~ s/\s+\}\n*$//s;
		$json .= ",\n";
		$output .= $json;
	}
	$output =~ s/,\n$/\n/;
	$output .= "}";
	#$self->logNote("output: ***$output***");
	open(OUT, ">$file") or die "Conf::Json::write    Can't open file: $file\n";
	print OUT $output;
	close(OUT) or die "Can't close file: $file\n";

	return 1;
};


method pretty ($file) {
	$self->logDebug("");
	$file = $self->inputfile() if not defined $file;
	my $contents = $self->fileContents($file);
	my $parser = $self->getJsonParser();
	my $object = $parser->decode($contents);
	my $json = $parser->pretty->indent->encode($object);
	$self->logDebug("json", $json);

	#### KEEP COPIES OF FILES IF backup DEFINED
	$self->makeBackup($file) if $self->backup();

	$self->toFile($file, $json);	
}

method unpretty ($file) {
	$file = $self->inputfile() if not defined $file;
	my $contents = $self->fileContents($file);
	my $parser = $self->getJsonParser();
	my $object = $parser->decode($contents);
	my $json = $parser->encode($object);

	#### KEEP COPIES OF FILES IF backup DEFINED
	$self->makeBackup($file) if $self->backup();

	$self->toFile($file, $json);	
}


method toFile ($file, $text) {
	$self->logDebug("file", $file);
	open(OUT, ">$file") or die "Can't open file: $file\n";
	print OUT $text;
	close(OUT) or die "Can't close file: $file\n";
}

method _getSections {
=head2

	SUBROUTINE 		_getSections
	
	PURPOSE
	
		RETURN ALL SECTIONS IN THE INPUTFILE
		
=cut
	return $self->sections() if defined $self->sections();
	my $sections = $self->read(undef);
	$self->sections($sections);

	return $sections;
}

method getKey ($key) {
	#$self->logDebug("");
	my $sections = $self->_getSections();
	$self->read(undef) if not defined $sections or not @$sections;
	return if not defined $sections;
	return $self->_getKey($key, $sections);
}

method _getKey ($key, $sections) {
=head2

	SUBROUTINE 		_getSection
	
	PURPOSE
	
		1. RETURN A SECTION WITH THE GIVEN KEY
		
=cut
	my $index = $self->_getKeyIndex($key, $sections);
	return if not defined $index;
	my $section = $$sections[$index];
	my @keys = keys %$section;
	my $section_key = $keys[0];
		
	return $section->{$section_key};
}


method getKeyIndex ($key) {
	#$self->logDebug("key", $key);
	my $sections = $self->_getSections();
	return $self->_getKeyIndex($key, $sections);
}

method _getKeyIndex ($key, $sections) {
	#$self->logDebug("key", $key);
	$sections = $self->_getSections() if not defined $sections;	
	#$self->logDebug("sections: ");
	#$self->logDebug(sprintf Dumper $sections);	
	return if not defined $sections;
	
	for ( my $i = 0; $i < @$sections; $i++ ) {
		my $section = $$sections[$i];
		my @keys = keys %$section;
		my $section_key = $keys[0];
		return $i if $section_key eq $key;
	}

	return;
}

method insertKey ($key, $value, $index) {
	#$self->logDebug("key", $key);
	#$self->logDebug("value", $value);
	#$self->logDebug("index", $index);
	$self->read(undef);
	my $sections = $self->_getSections();
	my $existingindex = $self->_getKeyIndex($key, $sections);
	#$self->logDebug("existingindex", $existingindex);	
	$index = $existingindex if not defined $index;
	
	### BY DEFAULT, PUT NEW KEY ENTRY AT TOP OF FILE
	$index = 0 if not defined $index;
	#$self->logDebug("index", $index);

	return if not defined $index;
	if ( defined $existingindex ) {
		$self->_removeKey($key, $existingindex, $sections);	
	}
	$self->_insertKey($key, $value, $index);
	my $outputfile = $self->outputfile() || $self->inputfile();
	$self->write($outputfile);
}

method _insertKey ($key, $value, $index) {
    #$self->logDebug("key", $key);
    #$self->logDebug("value", $value);
    #$self->logDebug("index", $index);
	my $sections = $self->_getSections();	
	$index = scalar(@$sections) if not defined $index;
	#$self->logDebug("BEFORE splice sections: ");
	#$self->logDebug(sprintf Dumper $sections);
    splice @{$sections}, $index, 0, { $key => $value };
	$self->sections($sections);
	#$self->logDebug("AFTER splice sections: ");
	#$self->logDebug(sprintf Dumper $sections);
	
	return $sections;
}

method removeKey ($key) {
	my $inputfile = $self->inputfile();
	$self->read($inputfile);
	my $sections = $self->_getSections();
	my $index = $self->_getKeyIndex($key, $sections);
	$self->_removeKey($key, $index, $sections);
	my $outputfile = $self->outputfile();
	$outputfile = $self->inputfile() if not defined $outputfile;
	$self->write($outputfile);
}

method _removeKey ($key, $index, $sections) {
	#$self->logDebug("key", $key);
	return if not defined $index;
	splice (@{$sections}, $index, 1);
	$self->sections($sections);
#
#	my $sections = $self->_getSections();	
#	my $currentindex = $self->_getKeyIndex($key, $sections);
#    splice (@{$sections}, $currentindex, 1) if defined $currentindex;
#	$index = $currentindex if not defined $index;

}

method hasKey ($key) {
	return $self->_getKeyIndex($key);	
	return 0;
}

##################			HOUSEKEEPING SUBROUTINES			################
method getJsonParser {
	return $self->jsonparser() if defined $self->jsonparser();
	my $jsonparser = JSON->new();
	$self->jsonparser($jsonparser);

	return $self->jsonparser();
}

method parseJson ($inputfile) {
	$self->logNote("inputfile", $inputfile);
	my $contents = $self->fileContents($inputfile);
	#$self->logDebug("contents", $contents);
	my $parser = $self->getJsonParser();
	return $parser->decode($contents);
}

method fileContents ($filename) {
	$self->logNote("filename", $filename);
	open(FILE, $filename) or die "Can't open filename: $filename\n";
	my $oldsep = $/;
	$/ = undef;
	my $contents = <FILE>;
	close(FILE);
	$/ = $oldsep;
	
	return $contents;
}

method dump { 
	$self->logDebug("Doing StarCluster::dump(@_)");
    require Data::Dumper;
    $Data::Dumper::Maxdepth = shift if @_;
    print Data::Dumper::Dumper $self;
}


=head1 LICENCE

This code is released under the GPL, a copy of which should
be provided with the code.

=end pod

=cut



}
